File: DevEnvironment/Command/src/main/java/command/core/ParameterSetParser.java
package command.core;

import base.core.Conditions;
import base.core.ParsingResult;
import base.interfaces.IParser;
import command.interfaces.IIndexedParameterMetadata;
import command.interfaces.IInputParameters;
import command.interfaces.INamedParameterMetadata;
import command.interfaces.IParameter;
import command.interfaces.IParameterSet;
import command.interfaces.IParameterSetMetadata;
import base.interfaces.IParsingResult;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * The ParameterSetParser class implements a parser of a parameter-set.
 */
public final class ParameterSetParser implements IParser<IInputParameters, IParameterSet> {
    private final String commandName;
    private final int parameterSetIndex;
    private final IParameterSetMetadata metadata;

    /**
     * The ParameterSetParser constructor.
     */
    public ParameterSetParser(
        String commandName,
        int parameterSetIndex,
        IParameterSetMetadata metadata) {

        Conditions.validateNotNull(
            commandName,
            "The name of a command.");

        Conditions.validateNotNull(
            metadata,
            "The metadata of a parameter-set.");

        this.commandName = commandName;
        this.parameterSetIndex = parameterSetIndex;
        this.metadata = metadata;
    }

    /**
     * Parses a parameter-set.
     */
    @Override
    public IParsingResult<IParameterSet> parse(IInputParameters inputParameters) {
        Conditions.validateNotNull(
            inputParameters,
            "The input parameters.");

        IParsingResult<List<IParameter>> result = this.parseIndexedParameters(
            this.metadata.getIndexedParameters(),
            inputParameters.getIndexedParameters());

        if (result.failed()) {
            return ParsingResult.failureResult(result);
        }

        List<IParameter> parameters = new ArrayList<>(result.getResult());

        result = this.parseNamedParameters(
            this.metadata.getNamedParameters(),
            inputParameters.getNamedParameters());

        if (result.failed()) {
            return ParsingResult.failureResult(result);
        }

        parameters.addAll(result.getResult());

        IParameterSet parameterSet = new ParameterSet(
            this.parameterSetIndex,
            this.metadata,
            parameters);

        return ParsingResult.successfulResult(parameterSet);
    }

    /**
     * Parses indexed parameters of a parameter-set.
     */
    private IParsingResult<List<IParameter>> parseIndexedParameters(
        List<IIndexedParameterMetadata> parametersMetadata,
        List<String> indexedParameters) {

        if (indexedParameters.size() != parametersMetadata.size()) {
            String errorMessage =
                "The input parameters for ParameterSet " + this.parameterSetIndex +
                " of command: " + this.commandName +
                " includes " + indexedParameters.size() + " indexed parameters," +
                " while it's metadata includes " + parametersMetadata.size() +
                "; Exactly " + parametersMetadata.size() + " are required.";

            return ParsingResult.failureResult(errorMessage);
        }

        List<IParameter> parameters = new ArrayList<>();

        int size = indexedParameters.size();
        for (int i = 0; i < size; ++i) {
            String parameterValue = indexedParameters.get(i);
            IIndexedParameterMetadata parameterMetadata = parametersMetadata.get(i);

            IParameter parameter = new Parameter(parameterMetadata, parameterValue);
            parameters.add(parameter);
        }

        return ParsingResult.successfulResult(parameters);
    }

    /**
     * Parses named parameters of a parameter-set.
     */
    private IParsingResult<List<IParameter>> parseNamedParameters(
        List<INamedParameterMetadata> parametersMetadata,
        Map<String, String> namedParameters) {

        List<IParameter> parameters = new ArrayList<>();

        for (INamedParameterMetadata parameterMetadata : parametersMetadata) {
            String parameterValue;

            if (namedParameters.containsKey(parameterMetadata.getShortName())) {
                parameterValue = namedParameters.get(parameterMetadata.getShortName());
            }
            else if (namedParameters.containsKey(parameterMetadata.getLongName())) {
                parameterValue = namedParameters.get(parameterMetadata.getLongName());
            }
            else if (parameterMetadata.optional()) {
                parameterValue = parameterMetadata.getDefaultValue();
            }
            else {
                String errorMessage =
                    "The input parameters for ParameterSet " + this.parameterSetIndex +
                    " of command: " + this.commandName +
                    " missing the following named parameter: " + parameterMetadata;

                return ParsingResult.failureResult(errorMessage);
            }

            IParameter parameter = new Parameter(parameterMetadata, parameterValue);
            parameters.add(parameter);
        }

        return ParsingResult.successfulResult(parameters);
    }
}

File: SystemCore/Command/core/ParameterSetParser.cpp
#include "ParameterSetParser.h"
#include "IInputParameters.h"
#include "ParameterSet.h"
#include "ParsingResult.h"
#include "SmartPointers.h"

using namespace command;

/**
 * Creates a new parameter set parser.
 */
IParameterSetParserSharedPtr ParameterSetParser::Make(
    const std::wstring& commandName,
    int parameterSetIndex,
    IParameterSetMetadataSharedPtr metadata)
{
    return std::make_shared<ParameterSetParser>(
        commandName,
        parameterSetIndex,
        metadata);
}

/**
 * The ParameterSetParser constructor.
 */
ParameterSetParser::ParameterSetParser(
    const std::wstring& commandName,
    int parameterSetIndex,
    IParameterSetMetadataSharedPtr metadata) :
    m_commandName(commandName),
    m_parameterSetIndex(parameterSetIndex)
{
    base::SmartPointers::Validate(metadata);
    m_metadata = metadata;
}

/**
 * The ParameterSetParser destructor.
 */
ParameterSetParser::~ParameterSetParser()
{
}

/**
 * Parses a parameter-set.
 */
IParsingResultSharedPtr<IParameterSetSharedPtr> ParameterSetParser::Parse(const IInputParameters& inputParameters)
{
    //
    // Parse indexed parameters...
    //
    std::vector<IParameterMetadataSharedPtr> indexedParametersMetadata;
    m_metadata->GetIndexedParameters(indexedParametersMetadata);

    const IInputParameters::IndexedParameters& indexedParameters = inputParameters.GetIndexedParameters();

    IParsingResultSharedPtr<std::vector<IParameterSharedPtr>> indexedParametersResult = ParseIndexedParameters(
        indexedParametersMetadata,
        indexedParameters);

    if (indexedParametersResult->Failed()) {
        return ParsingResult<IParameterSetSharedPtr>::FailureResult(indexedParametersResult);
    }

    //
    // Parse named parameters...
    //
    std::vector<IParameterMetadataSharedPtr> namedParametersMetadata;
    m_metadata->GetNamedParameters(namedParametersMetadata);

    const IInputParameters::NamedParameters& namedParameters = inputParameters.GetNamedParameters();

    IParsingResultSharedPtr<std::vector<IParameterSharedPtr>> namedParametersResult = ParseNamedParameters(
        namedParametersMetadata,
        namedParameters);

    if (namedParametersResult->Failed()) {
        return ParsingResult<IParameterSetSharedPtr>::FailureResult(namedParametersResult);
    }

    //
    // Create resultant parameter set...
    //
    std::vector<IParameterSharedPtr> parameters(
        indexedParametersResult->GetResult().begin(),
        indexedParametersResult->GetResult().end());

    parameters.insert(
        parameters.end(),
        namedParametersResult->GetResult().begin(),
        namedParametersResult->GetResult().end());

    IParameterSetSharedPtr parameterSet = ParameterSet::Make(
        m_parameterSetIndex,
        m_metadata,
        parameters);

    return ParsingResult<IParameterSetSharedPtr>::SuccessfulResult(parameterSet);
}

/**
 * Parses indexed parameters of a parameter-set.
 */
IParsingResultSharedPtr<std::vector<IParameterSharedPtr>> ParameterSetParser::ParseIndexedParameters(
    const std::vector<IParameterMetadataSharedPtr>& parametersMetadata,
    const std::vector<std::wstring>& indexedParameters) {

    return nullptr;
}

/**
 * Parses named parameters of a parameter-set.
 */
IParsingResultSharedPtr<std::vector<IParameterSharedPtr>> ParameterSetParser::ParseNamedParameters(
    const std::vector<IParameterMetadataSharedPtr>& parametersMetadata,
    const std::map<std::wstring, std::wstring>& namedParameters) {

    return nullptr;
}
